File: AstroPhysics/Alice/src/memory_store.py
from __future__ import annotations

import re
import sqlite3
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path


@dataclass(frozen=True)
class MemoryItem:
    id: int
    content: str
    category: str
    created_at: str
    use_count: int


def _normalize_text(value: str) -> str:
    lowered = value.strip().lower()
    lowered = re.sub(r"[^\w\s]", " ", lowered)
    lowered = re.sub(r"\s+", " ", lowered)
    return lowered.strip()


def _tokenize(value: str) -> list[str]:
    return [token for token in _normalize_text(value).split() if len(token) >= 3]


class MemoryStore:
    def __init__(self, db_path: Path) -> None:
        self._db_path = db_path.expanduser().resolve()
        self._db_path.parent.mkdir(parents=True, exist_ok=True)
        self._init_schema()

    @property
    def db_path(self) -> Path:
        return self._db_path

    def _connect(self) -> sqlite3.Connection:
        connection = sqlite3.connect(self._db_path)
        connection.row_factory = sqlite3.Row
        return connection

    def _init_schema(self) -> None:
        with self._connect() as connection:
            connection.execute(
                """
                CREATE TABLE IF NOT EXISTS memories (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    content TEXT NOT NULL,
                    normalized TEXT NOT NULL UNIQUE,
                    category TEXT NOT NULL DEFAULT 'general',
                    created_at TEXT NOT NULL,
                    last_used_at TEXT,
                    use_count INTEGER NOT NULL DEFAULT 0
                )
                """
            )
            connection.commit()

    def count(self) -> int:
        with self._connect() as connection:
            row = connection.execute("SELECT COUNT(*) AS n FROM memories").fetchone()
            return int(row["n"]) if row is not None else 0

    def add(self, content: str, *, category: str = "general") -> bool:
        cleaned = " ".join(content.strip().split())
        if not cleaned:
            return False
        normalized = _normalize_text(cleaned)
        if not normalized:
            return False

        timestamp = datetime.now().isoformat(timespec="seconds")
        with self._connect() as connection:
            try:
                connection.execute(
                    """
                    INSERT INTO memories (content, normalized, category, created_at, last_used_at, use_count)
                    VALUES (?, ?, ?, ?, ?, ?)
                    """,
                    (cleaned, normalized, category, timestamp, timestamp, 0),
                )
                connection.commit()
                return True
            except sqlite3.IntegrityError:
                return False

    def _touch(self, memory_ids: list[int]) -> None:
        if not memory_ids:
            return
        timestamp = datetime.now().isoformat(timespec="seconds")
        placeholders = ",".join("?" for _ in memory_ids)
        with self._connect() as connection:
            connection.execute(
                f"""
                UPDATE memories
                SET use_count = use_count + 1,
                    last_used_at = ?
                WHERE id IN ({placeholders})
                """,
                [timestamp, *memory_ids],
            )
            connection.commit()

    def recent(self, *, limit: int = 5) -> list[MemoryItem]:
        with self._connect() as connection:
            rows = connection.execute(
                """
                SELECT id, content, category, created_at, use_count
                FROM memories
                ORDER BY datetime(created_at) DESC
                LIMIT ?
                """,
                (limit,),
            ).fetchall()
        return [
            MemoryItem(
                id=int(row["id"]),
                content=str(row["content"]),
                category=str(row["category"]),
                created_at=str(row["created_at"]),
                use_count=int(row["use_count"]),
            )
            for row in rows
        ]

    def search(self, query: str, *, limit: int = 5) -> list[MemoryItem]:
        trimmed = query.strip()
        if not trimmed:
            return self.recent(limit=limit)

        tokens = _tokenize(trimmed)
        with self._connect() as connection:
            rows = connection.execute(
                """
                SELECT id, content, category, created_at, use_count
                FROM memories
                ORDER BY datetime(created_at) DESC
                LIMIT 300
                """
            ).fetchall()

        scored: list[tuple[int, MemoryItem]] = []
        query_norm = _normalize_text(trimmed)
        for row in rows:
            content = str(row["content"])
            normalized = _normalize_text(content)
            score = 0
            if query_norm and query_norm in normalized:
                score += 8
            for token in tokens:
                if token in normalized:
                    score += 2
            if score == 0:
                continue
            scored.append(
                (
                    score + min(int(row["use_count"]), 5),
                    MemoryItem(
                        id=int(row["id"]),
                        content=content,
                        category=str(row["category"]),
                        created_at=str(row["created_at"]),
                        use_count=int(row["use_count"]),
                    ),
                )
            )

        scored.sort(key=lambda pair: pair[0], reverse=True)
        selected = [item for _, item in scored[:limit]]
        self._touch([item.id for item in selected])
        return selected

File: AstroPhysics/Alice/src/memory_store.cpp
#include "alice/memory_store.hpp"

#include "alice/string_utils.hpp"

#include <algorithm>
#include <fstream>
#include <sstream>

namespace alice {

MemoryStore::MemoryStore(std::filesystem::path db_path) : db_path_(std::move(db_path)) {
    db_path_ = std::filesystem::weakly_canonical(db_path_.is_absolute() ? db_path_ : std::filesystem::current_path() / db_path_);
    std::filesystem::create_directories(db_path_.parent_path());
    load();
}

const std::filesystem::path& MemoryStore::db_path() const {
    return db_path_;
}

int MemoryStore::count() const {
    return static_cast<int>(items_.size());
}

std::string MemoryStore::escape(const std::string& value) {
    std::string out;
    out.reserve(value.size());
    for (char c : value) {
        if (c == '\\') {
            out += "\\\\";
        } else if (c == '\t') {
            out += "\\t";
        } else if (c == '\n') {
            out += "\\n";
        } else {
            out.push_back(c);
        }
    }
    return out;
}

std::string MemoryStore::unescape(const std::string& value) {
    std::string out;
    out.reserve(value.size());
    for (std::size_t i = 0; i < value.size(); ++i) {
        if (value[i] != '\\' || i + 1 >= value.size()) {
            out.push_back(value[i]);
            continue;
        }
        const char next = value[i + 1];
        if (next == 't') {
            out.push_back('\t');
            ++i;
        } else if (next == 'n') {
            out.push_back('\n');
            ++i;
        } else {
            out.push_back(next);
            ++i;
        }
    }
    return out;
}

void MemoryStore::load() {
    items_.clear();
    next_id_ = 1;

    std::ifstream file(db_path_);
    if (!file.is_open()) {
        return;
    }

    std::string line;
    while (std::getline(file, line)) {
        if (trim(line).empty()) {
            continue;
        }

        std::vector<std::string> parts;
        std::string piece;
        std::istringstream row(line);
        while (std::getline(row, piece, '\t')) {
            parts.push_back(piece);
        }
        if (parts.size() < 5) {
            continue;
        }

        MemoryItem item;
        try {
            item.id = std::stoi(parts[0]);
            item.use_count = std::stoi(parts[1]);
        } catch (...) {
            continue;
        }
        item.created_at = unescape(parts[2]);
        item.category = unescape(parts[3]);
        item.content = unescape(parts[4]);

        items_.push_back(item);
        next_id_ = std::max(next_id_, item.id + 1);
    }
}

bool MemoryStore::save() const {
    std::ofstream file(db_path_, std::ios::trunc);
    if (!file.is_open()) {
        return false;
    }

    for (const auto& item : items_) {
        file << item.id << "\t" << item.use_count << "\t" << escape(item.created_at) << "\t" << escape(item.category)
             << "\t" << escape(item.content) << "\n";
    }
    return file.good();
}

bool MemoryStore::add(const std::string& content, const std::string& category) {
    const std::string cleaned = trim(content);
    if (cleaned.empty()) {
        return false;
    }
    const std::string normalized_new = normalize_text(cleaned);
    if (normalized_new.empty()) {
        return false;
    }

    for (const auto& item : items_) {
        if (normalize_text(item.content) == normalized_new) {
            return false;
        }
    }

    MemoryItem item;
    item.id = next_id_++;
    item.content = cleaned;
    item.category = category;
    item.created_at = now_iso8601();
    item.use_count = 0;

    items_.push_back(item);
    return save();
}

std::vector<MemoryItem> MemoryStore::recent(int limit) {
    if (limit <= 0) {
        limit = 5;
    }
    std::vector<MemoryItem> out;
    const int start = std::max(0, static_cast<int>(items_.size()) - limit);
    for (int i = static_cast<int>(items_.size()) - 1; i >= start; --i) {
        out.push_back(items_[static_cast<std::size_t>(i)]);
    }
    return out;
}

std::vector<MemoryItem> MemoryStore::search(const std::string& query, int limit) {
    if (limit <= 0) {
        limit = 5;
    }
    const std::string q = normalize_text(query);
    if (q.empty()) {
        return recent(limit);
    }
    const auto q_tokens = split_words(q);

    std::vector<std::pair<int, std::size_t>> scored;
    scored.reserve(items_.size());

    for (std::size_t i = 0; i < items_.size(); ++i) {
        const std::string normalized = normalize_text(items_[i].content);
        int score = 0;
        if (normalized.find(q) != std::string::npos) {
            score += 8;
        }
        for (const auto& token : q_tokens) {
            if (token.size() < 3) {
                continue;
            }
            if (normalized.find(token) != std::string::npos) {
                score += 2;
            }
        }
        score += std::min(items_[i].use_count, 5);
        if (score > 0) {
            scored.emplace_back(score, i);
        }
    }

    std::sort(scored.begin(), scored.end(), [](const auto& a, const auto& b) {
        return a.first > b.first;
    });

    std::vector<MemoryItem> result;
    for (int i = 0; i < static_cast<int>(scored.size()) && i < limit; ++i) {
        auto& item = items_[scored[static_cast<std::size_t>(i)].second];
        item.use_count += 1;
        result.push_back(item);
    }

    if (!result.empty()) {
        save();
    }
    return result;
}

}  // namespace alice
